File: packages/php/src/PalPluss.php
<?php

declare(strict_types=1);

namespace PalPluss;

use PalPluss\Http\HttpTransport;
use PalPluss\Modules\B2cModule;
use PalPluss\Modules\ChannelsModule;
use PalPluss\Modules\StkModule;
use PalPluss\Modules\TransactionsModule;
use PalPluss\Modules\WalletsModule;

/**
 * PalPluss API client.
 *
 * @example
 * ```php
 * $client = new PalPluss(apiKey: 'pk_live_...');
 * $result = $client->stkPush(amount: 500, phone: '254712345678');
 * echo $result['transactionId'];
 * ```
 */
final class PalPluss
{
    private readonly HttpTransport $transport;
    private readonly StkModule $stk;
    private readonly B2cModule $b2c;
    private readonly WalletsModule $wallets;
    private readonly TransactionsModule $transactions;
    private readonly ChannelsModule $channels;

    /**
     * @param string|null      $apiKey               Your PalPluss API key. Falls back to
     *                                               PALPLUSS_API_KEY environment variable.
     * @param float            $timeout              Request timeout in seconds. Default: 30.
     * @param bool             $autoRetryOnRateLimit Automatically retry on HTTP 429. Default: true.
     * @param int              $maxRetries           Maximum retry attempts. Default: 3.
     * @param HttpTransport|null $transport           @internal — for testing only.
     */
    public function __construct(
        ?string $apiKey = null,
        float $timeout = 30.0,
        bool $autoRetryOnRateLimit = true,
        int $maxRetries = 3,
        ?HttpTransport $transport = null,
    ) {
        if ($transport !== null) {
            $this->transport = $transport;
        } else {
            $resolvedKey = $apiKey ?? (getenv('PALPLUSS_API_KEY') ?: null);
            if ($resolvedKey === null || $resolvedKey === '') {
                throw new \InvalidArgumentException('PalPluss: api_key is required');
            }

            $baseUrl = getenv('PALPLUSS_BASE_URL') ?: 'https://api.palpluss.com/v1';

            $this->transport = new HttpTransport(
                apiKey: $resolvedKey,
                baseUrl: $baseUrl,
                timeout: $timeout,
                autoRetryOnRateLimit: $autoRetryOnRateLimit,
                maxRetries: $maxRetries,
            );
        }

        $this->stk = new StkModule($this->transport);
        $this->b2c = new B2cModule($this->transport);
        $this->wallets = new WalletsModule($this->transport);
        $this->transactions = new TransactionsModule($this->transport);
        $this->channels = new ChannelsModule($this->transport);
    }

    // ── STK Push ─────────────────────────────────────────────────────────────

    /**
     * Initiate an STK Push (Lipa na M-Pesa) payment.
     *
     * @return array<string, mixed>
     */
    public function stkPush(
        float $amount,
        string $phone,
        ?string $accountReference = null,
        ?string $transactionDesc = null,
        ?string $channelId = null,
        ?string $callbackUrl = null,
        ?string $credentialId = null,
    ): array {
        return $this->stk->initiate(
            amount: $amount,
            phone: $phone,
            accountReference: $accountReference,
            transactionDesc: $transactionDesc,
            channelId: $channelId,
            callbackUrl: $callbackUrl,
            credentialId: $credentialId,
        );
    }

    // ── B2C Payout ───────────────────────────────────────────────────────────

    /**
     * Send a B2C payout to a phone number.
     *
     * A UUID v4 idempotency key is auto-generated if not provided.
     *
     * @return array<string, mixed>
     */
    public function b2cPayout(
        float $amount,
        string $phone,
        ?string $currency = null,
        ?string $reference = null,
        ?string $description = null,
        ?string $channelId = null,
        ?string $credentialId = null,
        ?string $callbackUrl = null,
        ?string $idempotencyKey = null,
    ): array {
        return $this->b2c->payout(
            amount: $amount,
            phone: $phone,
            currency: $currency,
            reference: $reference,
            description: $description,
            channelId: $channelId,
            credentialId: $credentialId,
            callbackUrl: $callbackUrl,
            idempotencyKey: $idempotencyKey,
        );
    }

    // ── Service Wallet ────────────────────────────────────────────────────────

    /**
     * Retrieve the service wallet balance.
     *
     * @return array<string, mixed>
     */
    public function getServiceBalance(): array
    {
        return $this->wallets->serviceBalance();
    }

    /**
     * Top up the service wallet via STK Push.
     *
     * @return array<string, mixed>
     */
    public function serviceTopup(
        float $amount,
        string $phone,
        ?string $accountReference = null,
        ?string $transactionDesc = null,
        ?string $idempotencyKey = null,
    ): array {
        return $this->wallets->serviceTopup(
            amount: $amount,
            phone: $phone,
            accountReference: $accountReference,
            transactionDesc: $transactionDesc,
            idempotencyKey: $idempotencyKey,
        );
    }

    // ── Transactions ──────────────────────────────────────────────────────────

    /**
     * Retrieve a single transaction by ID.
     *
     * @return array<string, mixed>
     */
    public function getTransaction(string $transactionId): array
    {
        return $this->transactions->get($transactionId);
    }

    /**
     * List transactions with optional filters.
     *
     * @param  string|null $type  "STK" or "B2C"
     * @return array<string, mixed>
     */
    public function listTransactions(
        ?int $limit = null,
        ?string $cursor = null,
        ?string $status = null,
        ?string $type = null,
    ): array {
        return $this->transactions->list(
            limit: $limit,
            cursor: $cursor,
            status: $status,
            type: $type,
        );
    }

    // ── Payment Wallet Channels ───────────────────────────────────────────────

    /**
     * Create a payment wallet channel.
     *
     * @param  string $type "PAYBILL", "TILL", or "SHORTCODE"
     * @return array<string, mixed>
     */
    public function createChannel(
        string $type,
        string $shortcode,
        string $name,
        ?string $accountNumber = null,
        ?bool $isDefault = null,
    ): array {
        return $this->channels->create(
            type: $type,
            shortcode: $shortcode,
            name: $name,
            accountNumber: $accountNumber,
            isDefault: $isDefault,
        );
    }

    /**
     * Update a payment wallet channel.
     *
     * @return array<string, mixed>
     */
    public function updateChannel(
        string $channelId,
        ?string $type = null,
        ?string $shortcode = null,
        ?string $name = null,
        ?string $accountNumber = null,
        ?bool $isDefault = null,
    ): array {
        return $this->channels->update(
            channelId: $channelId,
            type: $type,
            shortcode: $shortcode,
            name: $name,
            accountNumber: $accountNumber,
            isDefault: $isDefault,
        );
    }

    /**
     * Delete a payment wallet channel.
     */
    public function deleteChannel(string $channelId): void
    {
        $this->channels->delete($channelId);
    }
}

File: packages/go/palpluss.go
// Package palpluss is the official Go SDK for the PalPluss payments API.
//
// Basic usage:
//
//	client, err := palpluss.New("pk_live_...")
//	if err != nil {
//	    log.Fatal(err)
//	}
//	result, err := client.StkPush(ctx, palpluss.StkPushParams{
//	    Amount: 500,
//	    Phone:  "254712345678",
//	})
package palpluss

import (
	"context"
	"crypto/rand"
	"fmt"
	"net/url"
	"os"
	"strconv"
	"time"
)

const defaultBaseURL = "https://api.palpluss.com/v1"

// Client is the PalPluss API client.
type Client struct {
	t *transport
}

type clientConfig struct {
	baseURL    string
	timeout    time.Duration
	autoRetry  bool
	maxRetries int
}

// Option configures the Client.
type Option func(*clientConfig)

// WithBaseURL sets a custom API base URL (overrides PALPLUSS_BASE_URL env var).
func WithBaseURL(u string) Option {
	return func(c *clientConfig) { c.baseURL = u }
}

// WithTimeout sets the HTTP request timeout (default: 30s).
func WithTimeout(d time.Duration) Option {
	return func(c *clientConfig) { c.timeout = d }
}

// WithAutoRetryOnRateLimit enables or disables automatic retries on HTTP 429
// (default: true).
func WithAutoRetryOnRateLimit(enabled bool) Option {
	return func(c *clientConfig) { c.autoRetry = enabled }
}

// WithMaxRetries sets the maximum number of retry attempts (default: 3).
func WithMaxRetries(n int) Option {
	return func(c *clientConfig) { c.maxRetries = n }
}

// New creates a new PalPluss client.
//
// apiKey may be empty, in which case the PALPLUSS_API_KEY environment variable
// is used. An error is returned if no key is available.
func New(apiKey string, opts ...Option) (*Client, error) {
	if apiKey == "" {
		apiKey = os.Getenv("PALPLUSS_API_KEY")
	}
	if apiKey == "" {
		return nil, fmt.Errorf("palpluss: api_key is required")
	}

	cfg := clientConfig{
		baseURL:    defaultBaseURL,
		timeout:    30 * time.Second,
		autoRetry:  true,
		maxRetries: 3,
	}
	if envURL := os.Getenv("PALPLUSS_BASE_URL"); envURL != "" {
		cfg.baseURL = envURL
	}
	for _, opt := range opts {
		opt(&cfg)
	}

	return &Client{
		t: newTransport(apiKey, cfg.baseURL, cfg.timeout, cfg.autoRetry, cfg.maxRetries),
	}, nil
}

// ── STK Push ──────────────────────────────────────────────────────────────────

// StkPush initiates an STK Push (Lipa na M-Pesa) payment request.
func (c *Client) StkPush(ctx context.Context, params StkPushParams) (*StkInitiateResponse, error) {
	body := map[string]any{
		"amount": params.Amount,
		"phone":  params.Phone,
	}
	if params.AccountReference != nil {
		body["accountReference"] = *params.AccountReference
	}
	if params.TransactionDesc != nil {
		body["transactionDesc"] = *params.TransactionDesc
	}
	if params.ChannelID != nil {
		body["channelId"] = *params.ChannelID
	}
	if params.CallbackURL != nil {
		body["callbackUrl"] = *params.CallbackURL
	}
	if params.CredentialID != nil {
		body["credential_id"] = *params.CredentialID
	}
	return call[StkInitiateResponse](ctx, c.t, "POST", "/payments/stk", body, nil, "")
}

// ── B2C Payout ────────────────────────────────────────────────────────────────

// B2cPayout sends a B2C payout to a phone number.
// A UUID v4 idempotency key is auto-generated if IdempotencyKey is nil.
func (c *Client) B2cPayout(ctx context.Context, params B2cPayoutParams) (*B2cPayoutResponse, error) {
	key := ""
	if params.IdempotencyKey != nil && *params.IdempotencyKey != "" {
		key = *params.IdempotencyKey
	} else {
		key = generateUUIDv4()
	}

	body := map[string]any{
		"amount": params.Amount,
		"phone":  params.Phone,
	}
	if params.Currency != nil {
		body["currency"] = *params.Currency
	}
	if params.Reference != nil {
		body["reference"] = *params.Reference
	}
	if params.Description != nil {
		body["description"] = *params.Description
	}
	if params.ChannelID != nil {
		body["channelId"] = *params.ChannelID
	}
	if params.CredentialID != nil {
		body["credential_id"] = *params.CredentialID
	}
	if params.CallbackURL != nil {
		body["callback_url"] = *params.CallbackURL
	}
	return call[B2cPayoutResponse](ctx, c.t, "POST", "/b2c/payouts", body, nil, key)
}

// ── Service Wallet ────────────────────────────────────────────────────────────

// GetServiceBalance retrieves the service wallet balance.
func (c *Client) GetServiceBalance(ctx context.Context) (*ServiceWalletBalance, error) {
	return call[ServiceWalletBalance](ctx, c.t, "GET", "/wallets/service/balance", nil, nil, "")
}

// ServiceTopup initiates an STK Push to top up the service wallet.
func (c *Client) ServiceTopup(ctx context.Context, params ServiceTopupParams) (*ServiceTopupResponse, error) {
	body := map[string]any{
		"amount": params.Amount,
		"phone":  params.Phone,
	}
	if params.AccountReference != nil {
		body["accountReference"] = *params.AccountReference
	}
	if params.TransactionDesc != nil {
		body["transactionDesc"] = *params.TransactionDesc
	}
	key := ""
	if params.IdempotencyKey != nil {
		key = *params.IdempotencyKey
	}
	return call[ServiceTopupResponse](ctx, c.t, "POST", "/wallets/service/topups", body, nil, key)
}

// ── Transactions ──────────────────────────────────────────────────────────────

// GetTransaction retrieves a single transaction by ID.
func (c *Client) GetTransaction(ctx context.Context, transactionID string) (*Transaction, error) {
	return call[Transaction](ctx, c.t, "GET", "/transactions/"+transactionID, nil, nil, "")
}

// ListTransactions returns a paginated list of transactions.
// All fields on ListTransactionsParams are optional.
func (c *Client) ListTransactions(ctx context.Context, params ListTransactionsParams) (*TransactionListResponse, error) {
	q := url.Values{}
	if params.Limit != nil {
		q.Set("limit", strconv.Itoa(*params.Limit))
	}
	if params.Cursor != nil {
		q.Set("cursor", *params.Cursor)
	}
	if params.Status != nil {
		q.Set("status", *params.Status)
	}
	if params.Type != nil {
		q.Set("type", *params.Type)
	}
	return call[TransactionListResponse](ctx, c.t, "GET", "/transactions", nil, q, "")
}

// ── Payment Wallet Channels ───────────────────────────────────────────────────

// CreateChannel creates a new payment wallet channel.
func (c *Client) CreateChannel(ctx context.Context, params CreateChannelParams) (*PaymentWalletChannel, error) {
	body := map[string]any{
		"type":      params.Type,
		"shortcode": params.Shortcode,
		"name":      params.Name,
	}
	if params.AccountNumber != nil {
		body["accountNumber"] = *params.AccountNumber
	}
	if params.IsDefault != nil {
		body["isDefault"] = *params.IsDefault
	}
	return call[PaymentWalletChannel](ctx, c.t, "POST", "/payment-wallet/channels", body, nil, "")
}

// UpdateChannel updates a payment wallet channel.
// Only non-nil fields in params are sent to the API.
func (c *Client) UpdateChannel(ctx context.Context, channelID string, params UpdateChannelParams) (*PaymentWalletChannel, error) {
	body := map[string]any{}
	if params.Type != nil {
		body["type"] = *params.Type
	}
	if params.Shortcode != nil {
		body["shortcode"] = *params.Shortcode
	}
	if params.Name != nil {
		body["name"] = *params.Name
	}
	if params.AccountNumber != nil {
		body["accountNumber"] = *params.AccountNumber
	}
	if params.IsDefault != nil {
		body["isDefault"] = *params.IsDefault
	}
	return call[PaymentWalletChannel](ctx, c.t, "PATCH", "/payment-wallet/channels/"+channelID, body, nil, "")
}

// DeleteChannel deletes a payment wallet channel.
func (c *Client) DeleteChannel(ctx context.Context, channelID string) error {
	_, err := c.t.do(ctx, "DELETE", "/payment-wallet/channels/"+channelID, nil, nil, "")
	return err
}

// ── Internal helpers ──────────────────────────────────────────────────────────

func generateUUIDv4() string {
	b := make([]byte, 16)
	_, _ = rand.Read(b)
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%08x-%04x-%04x-%04x-%012x",
		b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}
